### PR TITLE
Add Distribution, remove Histogram, Range and StdDev

### PR DESCRIPTION
**Pull request by songy23**
Per our discussion, `Histogram`, `Range` and `StdDev` will be replaced by a more general `Distribution` Aggregation. `Distribution` will have stats on sum, count, min, max, sum of squared deviations and histogram. `Distribution` Aggregation corresponds to StackDriver's data model Distribution: https://cloud.google.com/monitoring/api/ref_v3/rest/v3/TypedValue#Distribution.

Besides, `Distribution` for interval stats will only use N sub-intervals. The `Distribution` stats in the oldest sub-interval will be dropped.

This PR is built on PR #629, please review #629 before review this one. Might be easier to review by each commit.

### CODE REVIEW

File: core/src/main/java/io/opencensus/stats/Aggregation.java [hunk 1]
```diff
@@ -194,54 +122,52 @@ static Range create() {
      *
      * @return a new {@code Mean}.
      */
-    // TODO(songya): not v0.1, expose factory method later.
-    static Mean create() {
+    public static Mean create() {
```
Comment (dinooliva) [inline core/src/main/java/io/opencensus/stats/Aggregation.java]: I think Mean (and Sum and Count) are parameterless so you can create a static constant that you always return.

Comment (songy23) [reply core/src/main/java/io/opencensus/stats/Aggregation.java]: Sure, updated.

**Comment by sebright** [comment]
Should "not ready for merge" be removed now?

**Comment by songy23** [comment]
Sorry I forgot to remove that label. This PR is rebased and should be ready for merge now.

File: core/src/test/java/io/opencensus/stats/ViewDataTest.java [hunk 2]
```diff
@@ -197,22 +199,22 @@ public void preventStartTimeLaterThanEndTime() {
...
+  private static final Aggregation DISTRIBUTION = Distribution.create(BUCKET_BOUNDARIES);
+  
+  private static final ImmutableMap<List<TagValueString>, DistributionData> ENTRIES =
       ImmutableMap.of(
           Arrays.asList(V1, V2),
-          MeanData.create(0, 1),
+          DistributionData.create(1, 1, 1, 1, 0, 0, 1, 0),
           Arrays.asList(V10, V20),
-          MeanData.create(50, 2));
+          DistributionData.create(1, 1, 1, 1, 0, 0, 1, 0));
```
Comment (sebright) [inline core/src/test/java/io/opencensus/stats/ViewDataTest.java]: Should we use two different AggregationDatas, as before?

File: core/src/test/java/io/opencensus/stats/AggregationDataTest.java [hunk 3]
```diff
@@ -46,7 +46,7 @@
   @Test
   public void testCreateDistributionData() {
     DistributionData distributionData = DistributionData.create(1, 1, 1, 1, 0, 0, 1, 0);
-    assertThat(distributionData.getSum()).isWithin(TOLERANCE).of(1);
+    assertThat(distributionData.getMean()).isWithin(TOLERANCE).of(1);
     assertThat(distributionData.getCount()).isEqualTo(1);
     assertThat(distributionData.getMean()).isWithin(TOLERANCE).of(1);
```
Comment (sebright) [inline core/src/test/java/io/opencensus/stats/AggregationDataTest.java]: It looks like the getMean assertion is redundant.

File: core_impl/src/main/java/io/opencensus/implcore/stats/MutableAggregation.java [hunk 4]
```diff
@@ -355,76 +214,133 @@ long getCount() {
...
+      }
+      bucketCounts[bucketCounts.length - 1]++;
     }
 
+    // We don't compute fractional MutableDistribution, it's either whole or none.
     @Override
     void combine(MutableAggregation other, double fraction) {
-      throw new UnsupportedOperationException("Not supported in v0.1.");
+      checkArgument(other instanceof MutableDistribution, "MutableDistribution expected.");
+      if (fraction != 1) {
```
Comment (sebright) [inline core_impl/src/main/java/io/opencensus/implcore/stats/MutableAggregation.java]: This comparison should use a tolerance.

File: core_impl/src/main/java/io/opencensus/implcore/stats/MutableAggregation.java [hunk 5]
```diff
@@ -355,76 +214,133 @@ long getCount() {
...
+
+      MutableDistribution mutableDistribution = (MutableDistribution) other;
+      checkArgument(this.bucketBoundaries.equals(mutableDistribution.bucketBoundaries),
+          "Bucket boundaries should match.");
+
+      this.sum += mutableDistribution.sum;
+      this.count += mutableDistribution.count;
+      this.mean = this.sum / this.count;
+      this.min = mutableDistribution.min < this.min ? mutableDistribution.min : this.min;
+      this.max = mutableDistribution.max > this.max ? mutableDistribution.max : this.max;
```
Comment (sebright) [inline core_impl/src/main/java/io/opencensus/implcore/stats/MutableAggregation.java]: Suggestion: Setting min and max might be simpler with an if statement, as in add(double).

File: core_impl/src/main/java/io/opencensus/implcore/stats/MutableAggregation.java [hunk 6]
```diff
@@ -355,76 +214,133 @@ long getCount() {
...
+      // TODO(songya): how to calculate combined sum of squared deviations?
```
Comment (sebright) [inline core_impl/src/main/java/io/opencensus/implcore/stats/MutableAggregation.java]: Could you please create an issue for this, so that we don't forget?

File: core_impl/src/test/java/io/opencensus/implcore/stats/ViewManagerImplTest.java [hunk 7]
```diff
@@ -270,7 +271,7 @@ public void testRecordInterval() {
```
Comment (sebright) [inline core_impl/src/test/java/io/opencensus/implcore/stats/ViewManagerImplTest.java]: Should the two interval view tests also test with `DISTRIBUTION`, once sum of squared deviations is fully implemented?  I think there should be a TODO.

File: core_impl/src/main/java/io/opencensus/implcore/stats/MutableAggregation.java [hunk 8]
```diff
@@ -223,10 +223,9 @@ long getCount() {
   /** Calculate distribution stats on aggregated {@code MeasureValue}s. */
   static final class MutableDistribution extends MutableAggregation {
 
-    private double sum = 0.0;
+    private double mean = 0.0;
```
Comment (sebright) [inline core_impl/src/main/java/io/opencensus/implcore/stats/MutableAggregation.java]: What is the reason for this change?  I thought that the math was simpler when we had a sum field.

File: core/src/main/java/io/opencensus/stats/AggregationData.java [hunk 9]
```diff
@@ -128,126 +122,95 @@ public static CountData create(long count) {
...
+     * @param min min value.
+     * @param max max value.
+     * @param sumOfSquaredDeviations sum of squared deviations.
+     * @param bucketCounts histogram bucket counts.
+     * @return a {@code DistributionData}.
      */
-    public static RangeData create(double min, double max) {
+    public static DistributionData create(
+        double mean, long count, double min, double max, double sumOfSquaredDeviations,
+        long... bucketCounts) {
```
[the remark this answers is not in the document]
Comment (songy23) [reply core/src/main/java/io/opencensus/stats/AggregationData.java]: Make sense, done

File: core/src/test/java/io/opencensus/stats/AggregationDataTest.java [hunk 10]
```diff
@@ -40,27 +38,35 @@
...
+
   @Rule
   public ExpectedException thrown = ExpectedException.none();
 
   @Test
-  public void testCreateHistogramData() {
-    HistogramData histogram = HistogramData.create(0, 0, 0);
-    assertThat(histogram.getBucketCounts()).containsExactly(0L, 0L, 0L).inOrder();
+  public void testCreateDistributionData() {
+    DistributionData distributionData = DistributionData.create(1, 1, 1, 1, 0, 0, 1, 0);
```
[the remark this answers is not in the document]
Comment (songy23) [reply core/src/test/java/io/opencensus/stats/AggregationDataTest.java]: Updated.

File: core/src/test/java/io/opencensus/stats/ViewDataTest.java (hunk 2 above)
Comment (songy23) [reply core/src/test/java/io/opencensus/stats/ViewDataTest.java]: Sure, updated.

File: core_impl/src/main/java/io/opencensus/implcore/stats/MutableAggregation.java (hunk 4 above)
Comment (songy23) [reply core_impl/src/main/java/io/opencensus/implcore/stats/MutableAggregation.java]: Nice catch, updated.

File: core/src/test/java/io/opencensus/stats/AggregationDataTest.java (hunk 3 above)
Comment (songy23) [reply core/src/test/java/io/opencensus/stats/AggregationDataTest.java]: Good catch, removed.

File: core_impl/src/main/java/io/opencensus/implcore/stats/MutableAggregation.java (hunk 5 above)
Comment (songy23) [reply core_impl/src/main/java/io/opencensus/implcore/stats/MutableAggregation.java]: Done.

File: core_impl/src/main/java/io/opencensus/implcore/stats/MutableAggregation.java (hunk 6 above)
Comment (songy23) [reply core_impl/src/main/java/io/opencensus/implcore/stats/MutableAggregation.java]: Created https://github.com/census-instrumentation/opencensus-java/issues/649 for this

File: core_impl/src/main/java/io/opencensus/implcore/stats/MutableAggregation.java (hunk 8 above)
Comment (songy23) [reply core_impl/src/main/java/io/opencensus/implcore/stats/MutableAggregation.java]: Because to calculate sum of squared deviations, we still need to calculate mean first, and also two deltas from mean: https://github.com/census-instrumentation/opencensus-java/blob/master/core_impl/src/main/java/io/opencensus/implcore/stats/MutableAggregation.java#L394

In this case I think it's easier to store mean directly.

File: core_impl/src/test/java/io/opencensus/implcore/stats/ViewManagerImplTest.java (hunk 7 above)
Comment (songy23) [reply core_impl/src/test/java/io/opencensus/implcore/stats/ViewManagerImplTest.java]: Done.

File: core_impl/src/main/java/io/opencensus/implcore/stats/MutableAggregation.java (hunk 8 above)
Comment (sebright) [reply core_impl/src/main/java/io/opencensus/implcore/stats/MutableAggregation.java]: I only meant the change to remove the sum.  I thought it was simpler to store both the sum and the mean.

Comment (songy23) [reply core_impl/src/main/java/io/opencensus/implcore/stats/MutableAggregation.java]: Make sense, I've put back this field.

**Comment by songy23** [comment]
Thanks!
